haskell/actions.bzl

```diff
@@ -268,6 +270,28 @@ def _infer_rpaths(target, solibs):
 
   return r
 
+def _infer_deps(ctx):
```

Comment:
I'm concerned that these two functions are too easy to forget to call instead of reaching out for `ctx.attr.deps` etc, leading to subtle bugs down the road. Could this be done differently, e.g. by threading a provider all the way to the use site?

Reply:
:+1:
As another example where the logic feels a little fragile:
https://github.com/tweag/rules_haskell/pull/239/files/55b131bc543b781126aa11468bfbce7441842e14#diff-656a15c44401b2dbd8f5bfbc145d4f30R808

Rather than having two different ways to call `_haskell_library_impl` (either implicitly through a rule, or explicitly from within `_haskell_proto_aspect_impl`) I think it would be better to structure the code by defining three functions:

- `_haskell_library_impl()`, taking no arguments and used only in the rule definition of `haskell_library`
- `_haskell_proto_aspect_impl()` (taking no arguments)
- `compile_haskell_library(name, srcs, deps, {etc.})` , called by the above two.